opt: eliminate unnecessary NULL reordering of NOT NULL columns

mgartner opened the pull request:
In order to produce non-standard orderings of NULL values, i.e., when an
`ORDER BY` expression has an `ASC NULLS LAST` or `DESC NULLS FIRST`
clause, the optimizer plans "NULL reordering". Indexes never store
NULL values in these orderings, so the NULL reordering often results
in a full table scan and a sort.

However, observe that `ORDER BY col ASC NULLS FIRST` is equivalent to
`ORDER BY col ASC NULLS LAST` if `col` is known to not be NULL. The same
equivalence is true for `ORDER BY col DESC NULLS LAST` and
`ORDER BY col DESC NULLS FIRST`. Therefore, NULL reordering is unnecessary
if the ordering column is not nullable.

The `FoldIsNullProject` rule has been added which folds `x IS NOT NULL`
projections to `false` if `x` is known to be not `NULL`. This allows
other rules to fire, like `SimplifyRootOrdering`, which together
eliminate the NULL reordering steps, avoiding full table scans and
sorting operations.

Fixes #126675

Release note (performance improvement): The query optimizer now
generates more efficient plans for queries with clauses like
`ORDER BY col ASC NULLS LAST` and `ORDER BY col DESC NULLS FIRST` when
`col` is guaranteed to not be `NULL`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/126685)
<!-- Reviewable:end -->

DrewKimball reviewed (commented):
I like the idea, but I'm not sure this is the right place to make the change. What do you think of moving the logic to the optimizer, something like this?
```
diff --git a/pkg/sql/opt/norm/rules/project.opt b/pkg/sql/opt/norm/rules/project.opt
index 252eded2ad7..5ec9e7d2deb 100644
--- a/pkg/sql/opt/norm/rules/project.opt
+++ b/pkg/sql/opt/norm/rules/project.opt
@@ -330,3 +330,29 @@ $input
     (FoldJSONFieldAccess $projections $jsonCols $col $input)
     $passthrough
 )
+
+[FoldIsNullProject, Normalize]
+(Project
+    $input:*
+    $projections:[
+            ...
+            (ProjectionsItem
+                (Is (Variable $inCol) (Null))
+                $outCol
+            )
+            ...
+        ] &
+        (IsColNotNull $inCol $input)
+    $passthrough:*
+)
+=>
+(Project
+    $input
+    (ReplaceProjectionsItem # Note: this function doesn't exist yet.
+        $projections
+        $item
+        (ProjectionsItem (False) $outCol)
+    )
+    $passthrough
+)
```

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126685#-:-O0u3Dhy1TlMIMa1qR7g:bqtclf8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @mw5h)

___
*[`pkg/sql/opt/optbuilder/orderby.go` line 312 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126685#-O0u36fA2XNjC9TUNc9E:-O0u36fA2XNjC9TUNc9F:b-u3f75) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f2dfec82c2539f763ba9c3819ec7066be630569/pkg/sql/opt/optbuilder/orderby.go#L312)):*
> ```Go
> 	ord := tabID.ColumnOrdinal(sCol.id)
> 	col := b.factory.Metadata().Table(tabID).Column(ord)
> 	return !col.IsNullable()
> ```

I don't think this is always correct - what happens if the column passes through an outer-join, and is null-extended?


<!-- Sent from Reviewable.io -->

mgartner committed:
opt: eliminate unnecessary NULL reordering of NOT NULL columns

In order to produce non-standard orderings of NULL values, i.e., when an
`ORDER BY` expression has an `ASC NULLS LAST` or `DESC NULLS FIRST`
clause, the optimizer plans "NULL reordering". Indexes never store
NULL values in these orderings, so the NULL reordering often results
in a full table scan and a sort.

However, observe that `ORDER BY col ASC NULLS FIRST` is equivalent to
`ORDER BY col ASC NULLS LAST` if `col` is known to not be NULL. The same
equivalence is true for `ORDER BY col DESC NULLS LAST` and `ORDER BY col
DESC NULLS FIRST`. Therefore, NULL reordering is unnecessary if the
ordering column is not nullable.

The `FoldIsNullProject` rule has been added which folds `x IS NOT NULL`
projections to `false` if `x` is known to be not `NULL`. This allows
other rules to fire, like `SimplifyRootOrdering`, which together
eliminate the NULL reordering steps, avoiding full table scans and
sorting operations.

Fixes #126675

Release note (performance improvement): The query optimizer now
generates more efficient plans for queries with clauses like
`ORDER BY col ASC NULLS LAST` and `ORDER BY col DESC NULLS FIRST` when
`col` is guaranteed to not be `NULL`.

mgartner reviewed (commented):
>I like the idea, but I'm not sure this is the right place to make the change. What do you think of moving the logic to the optimizer, something like this?

You're absolutely right... What was I thinking?! :confused: 

A normalization rule is perfect for this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126685#-:-O0yXray4eA_9P3c4y57:b-m7fbaa)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @mw5h)

___
*[`pkg/sql/opt/optbuilder/orderby.go` line 312 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126685#-O0u36fA2XNjC9TUNc9E:-O0yZt2KEow6_jO2eO-W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f2dfec82c2539f763ba9c3819ec7066be630569/pkg/sql/opt/optbuilder/orderby.go#L312)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I don't think this is always correct - what happens if the column passes through an outer-join, and is null-extended?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
Nice! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I could imagine down the line we expand this logic to apply to other cases, maybe using other properties about the columns too. Seems like there's an opportunity to also do something similar for the `SimplifySelectFilter` path as well.

> Reviewed 5 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126685#-:-O0z4BTP0wkrPO4vcc_X:b-rw478p)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @mw5h)

___
*[`pkg/sql/opt/norm/project_funcs.go` line 907 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126685#-O0z30y463FrLtXJrv5U:-O0z30y5AUWFWrAO6OOK:bht59jx) ([raw file](https://github.com/cockroachdb/cockroach/blob/8278a5c0c2171580fe352e805ef045a7a96e497e/pkg/sql/opt/norm/project_funcs.go#L907)):*
> ```Go
> }
> 
> // FoldIsNullProjectionsItems folds all projections in the form "x IS NOT NULL"
> ```

[nit] Here and elsewhere, the rule as written actually matches `x IS NULL`, rather than `x IS NOT NULL`.
___
*[`pkg/sql/opt/norm/testdata/rules/project` line 1926 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126685#-O0z3jGDESzc-JygpBcn:-O0z3jGDESzc-JygpBco:bc81ijj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8278a5c0c2171580fe352e805ef045a7a96e497e/pkg/sql/opt/norm/testdata/rules/project#L1926)):*
> ```
> norm expect=FoldIsNullProject
> SELECT x IS NULL FROM a
> ----
> ```

Could we also do a case where the null is added by a join? Maybe like this:
```
SELECT a.x IS NULL, b.x IS NULL FROM a LEFT JOIN b ON a.x = b.x;
```


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> I could imagine down the line we expand this logic to apply to other cases, maybe using other properties about the columns too. Seems like there's an opportunity to also do something similar for the `SimplifySelectFilter` path as well.

:+1: 

TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126685#-:-O0z7uoP5gi7XM-WVGn1:b-672xj1)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @mw5h)

___
*[`pkg/sql/opt/norm/project_funcs.go` line 907 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126685#-O0z30y463FrLtXJrv5U:-O0z7wYZCRUggdalT3Or:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8278a5c0c2171580fe352e805ef045a7a96e497e/pkg/sql/opt/norm/project_funcs.go#L907)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Here and elsewhere, the rule as written actually matches `x IS NULL`, rather than `x IS NOT NULL`.
</blockquote></details>

Done.
___
*[`pkg/sql/opt/norm/testdata/rules/project` line 1926 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126685#-O0z3jGDESzc-JygpBcn:-O0z8wg27Vva-k3iJzw-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8278a5c0c2171580fe352e805ef045a7a96e497e/pkg/sql/opt/norm/testdata/rules/project#L1926)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Could we also do a case where the null is added by a join? Maybe like this:
```
SELECT a.x IS NULL, b.x IS NULL FROM a LEFT JOIN b ON a.x = b.x;
```
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
